doc/termnetd.1
.TH termnetd 1 01/08/00  GNU Terminal Server Daemon

.SH NAME
termnetd \- Terminal Server daemon

.SH SYNOPSIS
.B termnetd
[\-s port] [\-c configfile] [\-d level] [\-n] [\-p] [\-v]

.SH DESCRIPTION
This man page documents the
.BR termnetd
terminal daemon. This command allows telnet sessions to be established
with a unit's serial ports.
.PP
The following features are available:
.IP
A Control port for viewing and modifying the status of the ports
.IP
Telnet compatible.
.IP
Full device configuration.
.IP
Round robin port selection.
.IP
Multiple port configurations for each device.
.IP
Logs status & connection information to the system logs.
.IP
Live  setting of a device's baud rate and port configuration using the
.BR termnet
program.
.IP
Device  locking.  If    a   device is not available for a given TCP/IP
port,  the  connection  will be closed with no error message after the
connection  is  made.  Under  no  circumstances  will  2  simultaneous
connections be made to a single device.
.PP
The    devices  are  not  opened until a connection is established to
them.  Therefore, it is possible to still run Getty on the device. The
drawback  to  this is that another program can open a device and there
is no way for the daemon to tell that it is presently in use.
.SH OPTIONS
.TP
.I "\-c config\-file"
Set the configuration file to one other than the default of
.BR "/etc/termnetd.conf"
.TP
.I "\-d level"
Set the debug message level. Current levels are 1 - 5
.TP
.I \-n
Stops  the  daemon  from  forking  and  detaching  from the controlling
terminal. This is most useful for debugging purposes.
.TP
.I \-p
Set the promiscuous mode (Currently does nothing).
.TP
.I \-s port
Sets the IP port to listen to for the control port.
.TP
.I \-v
Sets the system logging to verbose.

.SH CONTROL PORT
The control port provides a simple interface for controlling the ports and 
viewing their status. To accomplish this, it has 7 commands. They are as 
follows:
.TP
.B enable port <TCP port expr>
Enable data transfer a port or device. This command is only valid for
ports with active network connections.
.TP
.B disable port <TCP port expr>
Disable data transfer a port or device. This command is only valid for
ports with active network connections.
.TP
.B disconnect port <TCP port expr>
Disconnects an existing network connection to a port or device. This command is 
only valid for ports with active network connections.
.TP
.B show port <TCP port expr>
Show the status of an individual port or all of the ports.
.TP
.B verbose
Toggle the displaying of a prompt and making the output a more user friendly.
Because this interface was primarily built for a another program to use. 
The default condition is to
.B not
have a prompt shown.
.TP
.B help
Display a short list and summary of commands.
.PP
The TCP port expressions are expected to be regular expressions to
match the ports with. For individual ports, their service name or the IP port
number may be given. To deal with multiple ports in one command, a regular 
expression should be used. To see or effect all ports, use the expression '.*'.
.PP
.SS Bug:
Sadly the the Control Port does not work with termnet! Use telnet to connect to it 
instead. The termnet program will just hang if you attempt to connect to 
the control port using it.

.SH CONFIGURATION
Configuration is accomplished through the file
.BR /etc/termnetd.conf .
A file with another name or path may be specified using the
.I \-c
option.  This  file consists of one or more entries with the following
format:
.IP
.BR "<IP port>:<state>:<device>:<termios options>;"
.PP
.SS FIELDS
.TP
.I "IP port"
Name  or number of the TCP/IP port to accept connections from for this
device.
.TP
.I state
Either 
.BR on
or
.BR off .
This  effectively enables or disables the port / device from accepting
connections.
.TP
.I device
The  name  of  the  device  to connect to. This must be in the form of
.BR /dev/<device> .
.TP
.I "termios options"
A  list  of  names  corresponding to the
.BR "termios options"
to be enabled
when the device is opened. These names must be capitalized and use the
same spelling as those specified in the 
.BR termios
man page.
.PP
Each  entry can use as many lines as needed to fit all of the required
termios  options.  Be  sure not to forget the semi-colon at the end of
the entry.
.PP
The  configuration  file  is  read  once at startup. The
.BR SIGHUP
signal  will  force the daemon to close any connections and re\-scan the
configuration file. 

.SS CONFIGURATION EXAMPLES
.BR "Example 1"
.PP
3000:on:/dev/cua1:B300 CLOCAL IGNBRK CRTSCTS CS8;
.br
3001:on:/dev/cua2:B300 CLOCAL IGNBRK CRTSCTS CS8;
.br
3002:on:/dev/cua3:B300 CLOCAL IGNBRK CRTSCTS CS8;
.PP
4000:on:/dev/cua1:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.br
4001:on:/dev/cua2:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.br
4002:on:/dev/cua3:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.PP
5000:on:/dev/cua1:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.br
5001:on:/dev/cua2:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.br
5002:on:/dev/cua3:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.PP
6000:on:/dev/cua1:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.br
6001:on:/dev/cua2:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.br
6002:on:/dev/cua3:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.PP
This configuration allows you to connect to three different modems,
.BR /dev/cua1 ,
.BR /dev/cua2
and
.BR /dev/cua3 ,
at  three different baud rates. If you connect to the TCP/IP socket at
port
.BR 300X ,
you will connect at 300 baud, while ports 
.BR 400X ,
.BR 500X 
and
.BR 500X
will set the baud rate to 2400, 9600 and 19200 respectively.
.PP
The
.I CLOCAL
option  allows  you  to communicate with a modem without carrier being
present. The
.I CRTSCTS
option  causes the CTS and RTS lines to be used for flow control. Note
that  as  far  as  I  can  tell,  SCO, AIX and OSF do not support this
option.
.PP
.BR "Example 2"
.PP
3000:on:/dev/cua3:B300 CLOCAL IGNBRK CRTSCTS CS8;
.br
3000:on:/dev/cua2:B300 CLOCAL IGNBRK CRTSCTS CS8;
.br
3000:on:/dev/cua1:B300 CLOCAL IGNBRK CRTSCTS CS8;
.PP
3001:on:/dev/cua3:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.br
3001:on:/dev/cua2:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.br
3001:on:/dev/cua1:B2400 CLOCAL IGNBRK CRTSCTS CS8;
.PP
3002:on:/dev/cua3:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.br
3002:on:/dev/cua2:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.br
3002:on:/dev/cua1:B9600 CLOCAL IGNBRK CRTSCTS CS8;
.PP
3003:on:/dev/cua3:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.br
3003:on:/dev/cua2:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.br
3003:on:/dev/cua1:B19200 CLOCAL IGNBRK CRTSCTS CS8;
.PP
This  configuration sets the ports up to be connected in a round robin
fashion. If multiple connections are made to the same socket, Ei. port
.BR 3000 ,
the  connection  will  be made to the first available device specified
for  the  port  number.  Note  that  the devices are listed in reverse
order.  This  is  because  the  internal  lists  to termnetd are built
backwards. Therefore, the first device to be connected to will be
.BR /dev/cua1 .

.SH FILES
/etc/termnetd.conf

.SH "SEE ALSO"
termnet(1), ttyd(1), termios(2)

.SH "KNOWN PROBLEMS"
Connections from telnet program that do no option negotiation may cause
problems. This has been experienced using telnet from AIX where the 
destination port is not 'telnet'. The AIX telnet program only sends telnet 
negotiation to connections to the port 'telnet'.
.SH AUTHOR
.PP
Joe Croft <joe@croftj.net>


CHANGES
05/22/02 Version 3.3
	Fixed spputc to let it write null chars out.
	Fixed configure script to use -f instead of -e
	Added proper includes to remove warnings.
	Brought the Licensing up todate.

05/22/02 Version 3.2
	Foralized the release.

09/01/01 Version 3.2.beta3
	Added patch provided by Vladimir Gurevich <vag@paulidav.org> for 
	building on Sun.
	Fixed a few compiler problems and warnings

07/22/01 Version 3.2.beta2
	Fixed Race condition causing ports left open by removing the dependency
	on SIGCHLD.

	Added ability to emulate 7bit Even parity data to the termnet client.

01/08/00 Version 3.1
	Moved the device opening in termnetd to the child processes to
	fix a race condition.

	Made small changes to the termnetd control port.

	Made changes to facilitate to trnVars.h building on the newer 
	Linux systems.


10/10/98 Version 3.0
	Added the -s option to termnetd to allow having a 'service port'.

	Added the ttyd command to allow use of remote serial devices by
	any program.

	Added uucp style serial device locking.

06/28/98 Version 2.05
	Modified termnetd to use sigaction instead of signal. This should
	make the server more robust in handling the death of children.

	Added a Control Port for checking and modifying the 
	status of the ports.

02/21/97 Version 2.03 -> 2.04
	Fixed nasty bug causing the server to die when DTR was pulled from
	a serial device

	Modified the build procedure to allow building for multiple platforms
	easier. Mostly used in the hope of using NFS at work.

	Added new platform support FreeBSD

	Fixed broken compilations on the AIX platform

11/15/96 Version 2.00 -> 2.03
	Mostly little bug fixes here and there
	(In other words I got lax keeping this log up to date, SORRY!)

03/16/96 Version 1.07 -> 2.00

	Moved  the  Special  telnet  options  to begin with 1. Because most
	telnet  daemons  do  not respond to the options, a 5 second timeout
	was  added.  If  this  timeout  expires,  it  is  assumed  that any
	responses  for  the  special termnet options have been received and
	normal execution will proceed.

	Added SCO DEC OSF the list of Unix flavors that the program runs on

	Completely overhauled the configuration files for termnetd. They
	now allow for using termios tags for the serial device configuration

02-02-96 Verion 1.06 -> 1.07

	Fixed the Telnet Option handshaking that caused telnet to go into
	LINEMODE when connecting to the server.

	Removed the -l and -r options from termnet and replaced them with
	the -e option to provide local echoing. Also changed the echo command
	to accept either on or off instead of local and remote. These changes
	were brought on by the first change.

	Added TELOPT TTYPE and TELOPT XDISPLOC option handshaking to termnet
	to provide a reasonable environment when connecting to telnet servers
	for login purposes.



termnetd/termnetd.c
/***************************************************************************
**    termnetd.c
**    Termnetd mainline code
**
** Copyright (C) 1996      Joseph Croft <joe@croftj.net>  
** 
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 1, or (at your option)
** any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
 ***************************************************************************/
#include "../sysdefs.h"
#include <errno.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <arpa/telnet.h>
#include "tnl.h"
#include "termnetd.h"

static int dbg = 0;

extern int optind, opterr;
extern char *optarg;

#ifdef OSF
extern int errno;
#endif

static portConf *pDevc;
static char confFile[1024] = "/etc/termnetd.conf";
static int Retry = 0;

#if defined (AIX) || defined(SunOS)
#define LOG_PERROR 0
#endif

void reloadConfig(int sig)
{
   syslog(LOG_INFO, "Reloading Configuration File %s", confFile);
   Restart = 1;
   Retry = 0;
}

void retryConnections(int sig)
{
   syslog(LOG_INFO, "Retrying Failed Connections");
   Restart = Retry = 1;
}

void ignoreSignal(int sig)
{
   syslog(LOG_INFO, "SIGHUP Received by parent!");
}

void enablePort(int sig)
{
   syslog(LOG_INFO, "SIGUSR2 Received by parent!");
   PortEnabled = 1;
}

void disablePort(int sig)
{
   syslog(LOG_INFO, "SIGUSR2 Received by parent!");
   PortEnabled = 0;
}

/*
** termnetd [options]
**
**    Options:
**
**       -c <file>   Use Alternate Configuration file 
**                   def: /etc/termnetd.conf
**
**       -d <dbg>    Set debug message 
**
**       -p          Permiscuous mode, no password needed to
**                   disable or disconnect ports
**
**       -n          Don't detach from the controlling terminal
**
**       -v          Verbose Message logging
*/

int main(int argc, char *argv[])
{
   int opt, cpid;
   struct sigaction sigact;

   while ((opt = getopt(argc, argv, "c:d:ps:nu:v")) != -1)
   {
      switch (opt)
      {
         case 'c':
            strcpy(confFile, optarg);
            break;

         case 'd':
            Debug = atoi(optarg);
            break;

         case 'p':
            Permiscuous = 1;
            break;

         case 'n':
            NoDetach = 1;
            break;

         case 'u':
            strcpy(PrivilegedUser, optarg);
            break;

         case 'v':
            Verbose = 1;
            break;

         case 's':
            strcpy(ControlPort, optarg);
            break;

         break;
      }
   }



   /*
   ** If they didn't ask us not to, detach from the controlling terminal
   ** and chdir to /etc.
   */
   if (NoDetach == 0)
   {
      openlog("termnetd", LOG_PID | LOG_CONS, LOG_DAEMON);
      if ((cpid = fork()) > 0)
         exit(0);
      else if (cpid < 0)
      {
         syslog(LOG_ERR, "Quiting!! Error Detaching from terminal:%m");
         exit(1);
      }
      else
      {
         chdir("/etc");
         close(0);
         close(1);
         close(2);
      }
   }
   else
      openlog("termnetd", LOG_PID | LOG_CONS | LOG_PERROR, LOG_DAEMON);

   if (loadConfig(confFile) < 0)
   {
      syslog(LOG_ERR, "Quiting!! Error opening config file %s:%m", confFile);
      exit(1);
   }
   memset(&sigact, 0, sizeof(sigact));
   sigact.sa_flags = SA_RESTART;

   sigact.sa_handler = reloadConfig;
   sigaction(SIGHUP, &sigact, NULL);
   sigact.sa_handler = retryConnections;
   sigaction(SIGALRM, &sigact, NULL);
   sigact.sa_handler = SIG_IGN;
   sigaction(SIGCHLD, &sigact, NULL);
   sigact.sa_handler = ignoreSignal;
   sigaction(SIGHUP, &sigact, NULL);
   for (;;)
   {
      Retry = 0;
      if (dbg) syslog(LOG_DEBUG, "main():Opening Sockets!");
      openSockets(Retry);
      if (dbg) syslog(LOG_DEBUG, "main():Calling Socket Select!");
      socketSelect();
      if (Retry == 0)
      {
         if (dbg) syslog(LOG_DEBUG, "main():Closing Sockets!");
         closeSockets();
      }
      if (dbg) syslog(LOG_DEBUG, "main():Loading Config file!");
      if (loadConfig(confFile) < 0)
      {
         syslog(LOG_ERR, "Quiting!! Error opening config file %s:%m", confFile);
         exit(1);
      }
      if (dbg) syslog(LOG_DEBUG, "main():Grabbing 60 winks!");
   }
#ifndef SCO
   exit(0);
#endif
}

void handleSpecialChars(int iVal, void *pArg1, void *pArg2)
{
   if (iVal == BREAK)
      SendBreak(pDevc->mfd);
}

void handleTelNetErrors(int iVal, void *pArg1, void *pArg2)
{
   if (iVal != 1)
      syslog(LOG_ERR, "Recieved Telnet error %d, rv = %d", iVal, (int)pArg1); 
   else
   {
      syslog(LOG_ERR, "Quit: Socket Closed");
      close(0);
      exit(1);
   }
}

void handleNewSubOptData(int iVal, void *pArg1, void *pArg2)
{
   char buff[80];

   if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():New SubOpt = %d", iVal);
   strcpy(buff, "");
   switch (iVal)
   {
      case TELOPT_BAUDRATE:
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Setting baudrate to %s",
                                     (const char *)pArg1);
         setBaudStr(pDevc, pArg1);
         setModem(pDevc->mfd, pDevc);
         strcpy(buff, pDevc->baud);
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Baudrate set to %s",
                                     pDevc->baud);
         break;

      case TELOPT_PORTSET:
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Setting port settings to %s",
                                     (const char *)pArg1);
         setPortStr(pDevc, pArg1);
         setModem(pDevc->mfd, pDevc);
         strcpy(buff, pDevc->port);
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Port Settings set to %s",
                                     pDevc->port);
         break;

      case TELOPT_DEVICE:
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Have Device request for %s",
                                     (const char *)pArg1);
         strcpy(buff, pDevc->devc);
         if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Port Settings set to %s",
                                     pDevc->port);
         break;
   }
   if (strlen(buff) > 0)
   {
      if (dbg) syslog(LOG_DEBUG, "handleNewSubOptData():Sending %s back", buff);
      tnlSendSubOption(iVal, buff);
   }
}

void hangup(int sig)
{
   syslog(LOG_INFO, "TTY Closed!, killing self");
   exit(0);
}

void termnetd(portConf *pDevice)
{
   int mfd, mError = 0, sflg = 0, cnt, rv;
   FILE *tnlin = NULL, *tnlout = NULL, *mfp = NULL;
   char commbuf[256];
   fd_set readFds, excpFds;
   int nSynch = 0;
   struct sigaction sigact;

   if (dbg) syslog(LOG_DEBUG, "telnetd():Enter");
   memset(&sigact, 0, sizeof(sigact));
   sigact.sa_flags = SA_RESTART;

   sigact.sa_handler = disablePort;
   sigaction(SIGUSR1, &sigact, NULL);
   sigact.sa_handler = enablePort;
   sigaction(SIGUSR2, &sigact, NULL);
   sigact.sa_handler = SIG_IGN;
   sigaction(SIGALRM, &sigact, NULL);
   sigact.sa_handler = SIG_IGN;
   sigaction(SIGCHLD, &sigact, NULL);

   pDevc = pDevice;
   mfd = openModem(pDevc);
   pDevc->mfd = mfd;

   PortEnabled = 1;
   if ((rv = tnlInit(NULL, NULL, &tnlin, &tnlout, 15)) != -1)
   {
      if (dbg) syslog(LOG_DEBUG, "telnetd():libtn initialized");

      /*
      ** Enable the proccessing of a few options & set the 
      ** Baud rate and port settings of the device
      */
      tnlSetCallBack(TNL_ISSUBOPTDATA_CB, handleNewSubOptData);
      tnlSetCallBack(TNL_SPECIALCHAR_CB, handleSpecialChars);
      tnlSetCallBack(TNL_ERROR_CB, handleTelNetErrors);
      tnlEnableOption(TELOPT_BAUDRATE);
      tnlEnableOption(TELOPT_PORTSET);
      tnlEnableOption(TELOPT_DEVICE);
      tnlEnableOption(TELOPT_ECHO);

      if (dbg) syslog(LOG_DEBUG, "telnetd():Initial Baudrate of |%s|",
                                 pDevice->baud);
      tnlSetSubOption(TELOPT_BAUDRATE, pDevice->baud);
      if (dbg) syslog(LOG_DEBUG, "telnetd():Initial Port parms of |%s|",
                                 pDevice->port);
      tnlSetSubOption(TELOPT_PORTSET, pDevice->port);
      tnlSetSubOption(TELOPT_DEVICE, pDevice->devc);

      tnlOfferOption(TELOPT_SGA, 1);
      tnlRequestOption(TELOPT_SGA, 1);
      if (dbg) syslog(LOG_DEBUG, "telnetd():Options offered");

      for (;;)
      {
         if (mfp == NULL)
         {
            if (dbg) syslog(LOG_DEBUG, "telnetd():Opening buffered I/O for device");
            if ((mfp = fdopen(mfd, "w")) == NULL)
            {
               syslog(LOG_ERR, "Error opening buffer I/O for device %s:%m",
                               pDevice->devc);
               closeModem(pDevc);
               close(0);
               exit(1);
            }
            if (dbg) syslog(LOG_DEBUG, "telnetd():Done opening buffered I/O for device");
         }

         FD_ZERO(&readFds);
         FD_ZERO(&excpFds);
         FD_SET(fileno(tnlin), &readFds);
         FD_SET(mfd, &readFds);

         if (dbg) syslog(LOG_DEBUG, "telnetd():Calling tnlSelect");
         if ((sflg = tnlSelect(&readFds, NULL, &excpFds, NULL)) >= 0)
         {
            if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Returned from  tnlSelect");
            if (mfd && FD_ISSET(mfd, &readFds))
            {
               if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Have data from Modem");
               if ((cnt = read(mfd, commbuf, sizeof(commbuf) - 1)) > 0)
               {
                  commbuf[cnt] = '\0';
                  if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Read %d bytes from modem: |%s|",
                                              cnt, commbuf);
                  if (PortEnabled) 
                  {
                     if ((cnt = write(fileno(tnlout), commbuf, cnt)) < 0)
                     {
                           syslog(LOG_ERR, "telnetd():Error Writing to network:%m");
                        syslog(LOG_INFO, "telnetd():Shutting down proccess");
                        fclose(mfp);
                        close(mfd);
                        close(0);
                        exit(0);
                     }
                     if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Wrote %d bytes to Network",
                                         cnt);
                  }
                  else
                     if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Chucked %d bytes from modem",
                                         cnt);
               }
               else if (cnt < 0 && errno != EAGAIN && errno != EINTR)
               {
                  syslog(LOG_ERR, "telnetd():Error Reading from modem:%m");
                  mError = 1;
                  mfd = 0;
                  mfp = NULL;
                  exit(1);
               }
            }

            if (FD_ISSET(fileno(tnlin), &readFds))
            {
               if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Have data from Network!");
               if ((cnt = read(fileno(tnlin), commbuf, sizeof(commbuf))) > 0)
               {
                  commbuf[cnt] = '\0';
                  if (dbg) syslog(LOG_DEBUG, "telnetd():Read %d bytes from Network: |%s|",
                                              cnt, commbuf);
                  if (PortEnabled) 
                  {
                     if ((cnt = write(mfd, commbuf, cnt)) < 0)
                     {
                           syslog(LOG_ERR, "telnetd():Error Writing to modem:%m");
                           syslog(LOG_INFO, "Shutting down proccess");
                           closeModem(pDevc);
                           fclose(mfp);
                           pDevc->mfd = -1;
                           close(0);
                           exit(0);
                     }
                     if (dbg) syslog(LOG_DEBUG, "telnetd():Wrote %d bytes to modem",
                                         cnt);
                  }
                  else
                     if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Chucked %d bytes from network",
                                         cnt);
               }
               else
               {
                  if (dbg > 2) syslog(LOG_ERR, "telnetd():Error reading network:%m");
                  if (dbg > 2) syslog(LOG_DEBUG, "telnetd():Error reading network:%m");
                  if (dbg > 2) syslog(LOG_INFO, "telnetd():Shutting down proccess");
                  closeModem(pDevc);
                  fclose(mfp);
                  pDevc->mfd = -1;
                  close(0);
                  exit(0);
               }
            }

            if (FD_ISSET(fileno(tnlin), &excpFds))
            {
               nSynch = 1;
            }
         }
         else
            syslog(LOG_ERR, "telnetd():Quiting!! Error from tnlSelect:%m");
      }
   }
   else
      syslog(LOG_ERR, "telnetd():Error connecting:m");
   closeModem(pDevc);
   fclose(mfp);
   pDevc->mfd = -1;
   close(0);
}     
